File: ThreemaFramework/DB/Entities/ImageMessageEntity.m
//  _____ _
// |_   _| |_  _ _ ___ ___ _ __  __ _
//   | | | ' \| '_/ -_) -_) '  \/ _` |_
//   |_| |_||_|_| \___\___|_|_|_\__,_(_)
//
// Threema iOS Client
// Copyright (c) 2012-2023 Threema GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License, version 3,
// as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

#import "ImageMessageEntity.h"
#import "NSString+Hex.h"
#import "UTIConverter.h"
#import "ThreemaFramework/ThreemaFramework-swift.h"

@implementation ImageMessageEntity

@dynamic encryptionKey;
@dynamic imageBlobId;
@dynamic imageNonce;
@dynamic imageSize;
@dynamic progress;

@dynamic image;
@dynamic thumbnail;


- (NSString*)logText {
    if ([self.image getCaption] != nil) {
        return [NSString stringWithFormat:@"%@ (%@) %@ %@", [BundleUtil localizedStringForKey:@"image"], [self blobFilename], [BundleUtil localizedStringForKey:@"caption"], [self.image getCaption]];
    }
    return [NSString stringWithFormat:@"%@ (%@)", [BundleUtil localizedStringForKey:@"image"], [self blobFilename]];
}

- (NSString*)previewText {
    return [BundleUtil localizedStringForKey:@"image"];
}

- (NSString *)quotePreviewText {
    NSString *caption = nil;
    if (self.image != nil) {
        caption = [self.image getCaption];
        if (caption != nil) {
            NSString *quotePreviewString = [[caption componentsSeparatedByCharactersInSet:[NSCharacterSet newlineCharacterSet]] componentsJoinedByString:@" "];
            caption = [quotePreviewString substringWithRange:NSMakeRange(0, MIN(caption.length, 200))];
        }
    }

    if (caption == nil || caption.length == 0) {
        caption = self.previewText;
    }
    return caption;
}

#pragma mark - Misc

#ifdef DEBUG
#else
- (NSString *)debugDescription
{
    return [NSString stringWithFormat:@"<%@: %p> %@ %@ %@ %@ %@ %@ %@ %@ %@ %@ %@ %@ %@ %@", [self class], self, @"image =", self.image.description, @"thumbnail =", self.thumbnail.description, @"imageBlobId =", @"***", @"imageNonce =", @"***", @"imageSize =", self.imageSize.description, @"progress =", self.progress.description, @"encryptionKey =", self.encryptionKey.description];
}
#endif

@end

File: ThreemaFramework/DB/Entities/ImageMessageEntity/ImageMessageEntity.swift
//  _____ _
// |_   _| |_  _ _ ___ ___ _ __  __ _
//   | | | ' \| '_/ -_) -_) '  \/ _` |_
//   |_| |_||_|_| \___\___|_|_|_\__,_(_)
//
// Threema iOS Client
// Copyright (c) 2024-2025 Threema GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License, version 3,
// as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

import CoreData
import Foundation

@objc(ImageMessageEntity)
public final class ImageMessageEntity: BaseMessageEntity {
    
    // MARK: Attributes

    @NSManaged public var encryptionKey: Data?
    // swiftformat:disable:next acronyms
    @NSManaged public var imageBlobId: Data?
    @NSManaged public var imageNonce: Data?
    @NSManaged public var imageSize: NSNumber?
    @NSManaged public var progress: NSNumber?

    // MARK: Relationships

    @NSManaged public var image: ImageDataEntity?
    @NSManaged public var thumbnail: ImageDataEntity?
    
    // MARK: Lifecycle
    
    /// Preferred initializer that ensures all non optional values are set
    /// - Parameters:
    ///   - context: `NSManagedObjectContext` to insert created entity into
    ///   - encryptionKey: Key the image data is encrypted with
    ///   - imageBlobID: BlobID of the image data
    ///   - imageNonce: Nonce of the image
    ///   - imageSize: Size of the image data
    ///   - progress: Progress
    ///   - image: `ImageDataEntity` of the image
    ///   - thumbnail: `ImageDataEntity` of the thumbnail
    public init(
        context: NSManagedObjectContext,
        encryptionKey: Data? = nil,
        imageBlobID: Data? = nil,
        imageNonce: Data? = nil,
        imageSize: NSNumber? = nil,
        progress: NSNumber? = nil,
        image: ImageDataEntity? = nil,
        thumbnail: ImageDataEntity? = nil
    ) {
        let entity = NSEntityDescription.entity(forEntityName: "ImageMessage", in: context)!
        super.init(entity: entity, insertInto: context)
        
        self.encryptionKey = encryptionKey
        // swiftformat:disable:next acronyms
        self.imageBlobId = imageBlobID
        self.imageNonce = imageNonce
        self.imageSize = imageSize
        self.progress = progress
        
        self.image = image
        self.thumbnail = thumbnail
    }
    
    @objc override private init(entity: NSEntityDescription, insertInto context: NSManagedObjectContext?) {
        super.init(entity: entity, insertInto: context)
    }
    
    @available(*, unavailable)
    public init() {
        fatalError("\(#function) not implemented")
    }
    
    @available(*, unavailable)
    public convenience init(context: NSManagedObjectContext) {
        fatalError("\(#function) not implemented")
    }
}
